### ISSUE TITLE
'recursive' #includes should not require includes/fplus on the include path

### ISSUE THREAD

**Issue by xtofl** [Dobiasd/FunctionalPlus#22]
E.g. composition.h contains an include

    #include "function_traits.h"

Which means that, when included from anywhere, the client project needs both `fplus_installation/includes` _and_ `fplus_installation/includes/fplus` on it's include path.

This makes adoption harder, and makes collisions with other libraries more probable.

Public headers should assume as little as possible from client code: they should probably include their neighbour headers as if specified by client code:

    #include "fplus/function_traits.h"

**Comment by Dobiasd** [comment]
Does the compiler not look into the directory of the current include file first? At least with G++, Clang and MSVC it seems to work without needing to point to `includes/fplus` as an additional include path.

**Comment by xtofl** [comment]
The `CMakeLists.txt` explicitly adds `includes/fplus`, that's why you don't see it in the current build.  You may want to create a 'fake' client project to guard against this.

And no: the compiler looks at the currently compiled _source file_ for neighbour include files.  (In other words, it adds the current compilation unit's path to the include path).

Take a look at Boost and Qt for example: their public headers only include stuff prefixed with `boost/` or `QtCore`.  (https://github.com/boostorg/range/blob/develop/include/boost/range.hpp as an example equivalent to `fplus.h`)

**Comment by Dobiasd** [comment]
Oh, I guess I now understand. `CMakeLists.txt` copies `include/fplus` and `include/fplus.h` into the systems include directory, so if one only includes `fplus.h` it should work, but including only specific parts like `#include <fplus/composition.h>` would be a problem, right?

I [changed the situation](https://github.com/Dobiasd/FunctionalPlus/commit/0cdf523244b5f8a4ba4f743bc81d061f12a31fba) according to your suggestion (boost style).

Also thanks for correcting my wrong include-path assumption. The [C++ standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf) verifies your statement.

**Comment by Dobiasd** [comment]
@xtofl: Just a small remark to our previous discussion, independently of the still meaningful change that resulted from it:

> the compiler looks at the currently compiled source file for neighbour include files.

Yes, but

> (In other words, it adds the current compilation unit's path to the include path).

seems to be incorrect. The current source file is not the compilation unit (cpp file) but the file doing the include. And this can be a header file too.

The following project does not compile:

`./main.cpp`:

    #include "foo/foo.h"
    #include <iostream>
    int main()
    {
        std::cout << x << std::endl;
    }

`./foo/foo.h`:

    #include "foo/bar/bar.h"

`./foo/bar/bar.h`:

    int x = 3;

Compiler error:

    g++ main.cpp 
    In file included from main.cpp:1:0:
    foo/foo.h:1:25: fatal error: foo/bar/bar.h: No such file or directory
     #include "foo/bar/bar.h"

It only works if the content from `foo.h` is changed to

    #include "bar/bar.h"

So the `bar.h` has to be included relative to `foo.h` not relative to `main.cpp`.

**Comment by xtofl** [comment]
## Thank you,  today I learned.

Indeed, searching for includes [appears implementation defined](http://en.cppreference.com/w/cpp/preprocessor/include)...  It is different [on msvc](https://msdn.microsoft.com/en-us/library/36k2cdd4.aspx) and [on gcc](https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html):

### Gcc:

> GCC looks for headers requested with #include "file" first in the directory containing the current file, then in the directories as specified by -iquote options, then in the same places it would have looked for a header requested with angle brackets. For example, if /usr/include/sys/stat.h contains #include "types.h", GCC looks for types.h first in /usr/include/sys, then in its usual search path.

### Msvc:

> The preprocessor searches for include files in this order:
In the same directory as the file that contains the #include statement.
In the directories of the currently opened include files, in the reverse order in which they were opened. The search begins in the directory of the parent include file and continues upward through the directories of any grandparent include files.
Along the path that's specified by each /I compiler option.
Along the paths that are specified by the INCLUDE environment variable.

What is common is that in the end, the preprocessor will look in the directories added via command line params -I (or /I), so I guess that for a library this was the right, portable, decision.

Also I learned that the angle bracket form is _intended_ to include implementation files, i.e. platform and compiler files, so we should probably use the qouted form.

> The intent of this syntax (`#include "xyz.h"`) is to search for the files that are not controlled by the implementation.

I thought I knew it all...

Thanks again!

**Comment by Dobiasd** [comment]
Ah, these descriptions are much easier to understand compared to the section in the C++ standard I linked to and obviously interpreted wrongly. :D

**Comment by Dobiasd** [comment]
@xtofl OK, the change with this issue consisted of two parts.
* **1**: move `fplus.h` into the `fplus` directory.
* **2**: Files like `filter.h` do no longer use for example `#include "maybe.h"` but `#include "fplus/maybe.h"` instead.

I am wondering if part **2** still makes sense in the light of our new understanding.

So Version **a** (`filter.h` uses `#include "maybe.h"`) allows the following use cases:
* **install** the fplus library to your system
* **download** fplus **and** **set** its include directory as an additional include path to the compiler.
* **download** fplus **and** **use an absolute path in the include** like `#include "/home/dobiasd/Downloads/FunctionalPlus/include/fplus/fplus.h"`.

Version **b** only supports the cases **"install"** and **"download and set"**. **"download and use an absolute path in the include"** only works for version **a**.
So is Version **a** really strictly better than version **b**? Should we switch back to version **a** or does it have any drawbacks I do not see yet?

**Comment by xtofl** [comment]
I see what you mean, and I admire the way you think about the installation use cases.

However, I think it bad practice to have code point to absolute paths: this makes the code work for one particular system at one particular point in time - something we should not advocate.  The search path for the compiler is a matter of configuration, not something we want the code to be burdened with.

Other programming systems also support libraries through an 'include-path' like settings (`PYTHON_PATH`, `CLASSPATH`, ...) and the C++ standard is even being extended with a full fletched Module system.

Besides, this may work in case the 'directory containing the current file' is searched, but will stop working from the point where we `#include` something from a 'sibling' directory, or want to evolve to a structure organized in more than one level of directories (like boost, qt, ace, ...) - imagine this:

    .../fplus/util/concat.h  (#include "string/concat.h", "array/concat.h", ...)
    .../fplus/string/concat.h

Now client code including `util/concat.h` will only compile under msvc.  Maybe not the best example, but I think it illustrates the point.

I think it's a good idea to support the 'system' and 'include-path' only.  This way you'll have two consistent ways the library will always work correctly: install to the system, or add to the include path.

In short: in my eyes version **a** is worse than version **b** since it supports an inferior programming style.

**Comment by Dobiasd** [comment]
Sure, absolute paths in an include are very bad practice. But I would not want to force the educational effect by not supporting it, if everything else would stay the same. You argument about future directory structures however convinces me. I now also think it is best to leave it as it is now. Thanks for helping me think all this through. :+1:
